azkaban-common/src/main/java/azkaban/project/ProjectManager.java

```diff
@@ -55,8 +56,8 @@
   private final boolean creatorDefaultPermissions;
   private final ConcurrentHashMap<Integer, Project> projectsById =
       new ConcurrentHashMap<>();
-  private final ConcurrentHashMap<String, Project> projectsByName =
-      new ConcurrentHashMap<>();
+  private final CaseInsensitiveConcurrentHashMap<Project> projectsByName =
```

Comment:
Have you considered using an existing library https://commons.apache.org/proper/commons-collections/apidocs/org/apache/commons/collections4/map/CaseInsensitiveMap.html ?

Reply:
Yes, I've considered the existing CaseInsensitiveMap as well. But it is not synchronized and is not thread-safe. Since the project cache will be accessed by multiple threads, it is necessary to keep it synchronized.

Reply:
As discussed offline, please check if external synchronization is already required.

Reply:
Thanks for bringing up this discussion. 
ConcurrentHashMap provides concurrent access for multiple threads to the project cache. ConcurrentHashMap is thread safe itself but multiple operations on the map might not be thread safe. (https://stackoverflow.com/questions/14947723/is-concurrenthashmap-totally-safe).
I found two places in the code, in azkaban.project.ProjectManager#createProject, when two threads first check the project cache, the project might not exist yet. One thread puts the new project to the cache first and then might be overwritten by the other thread. 
Also, when one thread is trying to get the project, another thread might call azkaban.project.ProjectManager#removeProject at the same time. The project returned in the first thread could be null. 
To solve the above potential issue, we could add synchronized block on the operations. But there are some tradeoffs. For multiple threads, synchronization will have some effects on the performance. Also, we might have to reevaluate all other cache and see if they require additional synchronization. That might need a separate PR if we want to solve the cache synchronization issue completely.
For this particular PR, I would prefer still keeping the map as concurrent unless we have clear evidence that the performance won't be affected if we use non-concurrent map instead.

Reply:
Thanks for the summary, as discussed offline. I think it's premature to optimize for performance here. I don't think the performance impact will be big enough to offset the benefits of more reliable/easier to reason about code. 

A separate PR is fine. 

Could you have it reviewed by other team members as well?

Reply:
There seems to be a bug in the existing code. Your change and investigation uncovered it. Thanks!

e.g. 
If two threads create projects with the same name at about the same time, it seems that they may both succeed and create two entries in the DB. What's in the cache is not deterministic, the projectsById cache may point to one and the projectsByName cache may point to another.
 
see
https://github.com/HappyRay/azkaban/blob/eb8450a55e03b7a5cbf3768373e5f091804ed880/azkaban-common/src/main/java/azkaban/project/ProjectManager.java#L270

The usage of synchronized keyword in this class is also inconsistent.

Granted that it should be very rare that such a race can happen and one may even argue that the resulting data corruption is acceptable given that it can be traced and corrected manually,  with a restart of the web server. You probably remember how costly and painful it is to debug timing related subtle bugs. I would rather spend time writing code that I have higher confidence that it is correct than debugging race conditions even if it is rare. And like other race condition bugs, it is hard to convince myself that I have covered all cases if we allow operations from multiple threads to interleave. That's what I mean by "easier to reason about" code.  

Given our access pattern and the chance of this race condition, I don't think I have a sufficiently strong argument to block your change. However, I would encourage you to seek other people's opinion and make your own decision. 

If you do decide to keep the current design, I'd suggest you add a comment to the code alerting readers of this code of this potential race condition and your reasoning to do it this way. 

As discussed, it should be possible to run a micro benchmark to evaluate how long the synchronization will add to the user-visible latency. It may be a beneficial exercise for you. It will help you build a better sense of how to estimate performance impact. Seek help from performance team if needed.

Reply:
Thanks @HappyRay! I totally agree with your point of fixing the current cache synchronization issue. Otherwise, it will ultimately bite us in the future. In this PR, I'll just focus on the consistency issue between project cache and DB. I'll submit another PR to address the cache synchronization issue and link to this one.

Reply:
Thanks. I am glad we are in agreement.